### ISSUE TITLE
Please expose _PyTime_t, _PyTime_FromTimeval, and _PyTime_AsSecondsDouble as public APIs

### ISSUE THREAD

**Issue by ryancdotorg** [python/cpython#110850]
# Feature or enhancement

### Proposal:

#106316 #106317 removed pytime.h, and I was using some things

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Linked PRs
* gh-112135
* gh-115215
* gh-115348
* gh-115713
* gh-115715
* gh-115719
* gh-115724
* gh-115734
* gh-115746
* gh-115751
* gh-115753
* gh-115973
* gh-116463
* gh-116606
* gh-118394
* gh-118552
* gh-118556

**Comment by hugovk** [comment]
Please could you say something about your use case? 

Why do you need these functions? 

Have you tried alternative APIs to achieve the same result? 

cc @vstinner

**Comment by ryancdotorg** [comment]
I have a C/C++ module that was using them which I wrote a couple years ago. It broke.

I actually dug into what I was doing, and it turns out it was basically just (error checking omitted):

```
_PyTime_t tp;
struct timeval tv;
get_timestamp(&tv, ...);
_PyTime_FromTimeval(&tp, &tv);
double ts = _PyTime_AsSecondsDouble(tp); 
```

and I was able to more sensibly replace it with

```
double tv_frac = tv.tv_usec / 1e6;
struct timeval tv;
get_timestamp(&tv, ...);
double tv_frac = tv.tv_usec / 1e6;
double ts = tv.tv_sec + tv_frac;
```

It seems I just didn't want to bother figuring out the safest way to do the conversion.

This can be closed.

**Comment by vstinner** [comment]
I'm open to make some functions public, but apparently here, there wasn't a strong need for that. Maybe it can wait. Thanks for the report anyway.

**Comment by scoder** [comment]
We actually expose that API in Cython's [`cpython/time.pxd`](https://github.com/cython/cython/blob/master/Cython/Includes/cpython/time.pxd). Please move it back to where it was.

**Comment by vstinner** [comment]
> We actually expose that API in Cython's [cpython/time.pxd](https://github.com/cython/cython/blob/master/Cython/Includes/cpython/time.pxd). Please move it back to where it was.

This file uses the following APIs:

* _PyTime_t
* _PyTime_GetSystemClock()
* _PyTime_AsSecondsDouble()

**Comment by scoder** [comment]
> This file uses the following APIs

Correct. It basically mimics the Python `time` module and reimplements a part of it in C.

I'd still recommend exposing the complete C-API. We are talking about a standard library module here, not a part of the core C-API, so it represents a natural API border already. We are not talking about internals here. Let's just keep that up and allow regular external use as before.

**Comment by vstinner** [comment]
I designed _PyTime_t API for Python internal usage, but apparently some people discovered it and decided to use it since it's convenient.

_PyTime_t API was designed to be able to write easily "deadline = time + timeout", but later I started to add "clamp" variants in code which **cannot** report errors to the caller. Clamp the result in the range [_PyTime_MIN; _PyTime_MAX] is better than having an undefined behavior.

One issue is that `_PyTime_t` range is smaller than time_t range:

```
// The _PyTime_t API supports a resolution of 1 nanosecond. The _PyTime_t type
// is signed to support negative timestamps. The supported range is around
// [-292.3 years; +292.3 years]. Using the Unix epoch (January 1st, 1970), the
// supported date range is around [1677-09-21; 2262-04-11].
```

64-bit time_t (standard even on 32-bit platforms) is way larger than that. **In practice**, it should not be an issue. But the datetime module cannot use `_PyTime_t` everywhere for example, since datetime max year is 9999, not year 2262.

I tried to make the `_PyTime_t` unit an implementation details and enforce _PyTime_FromNanoseconds() or _PyTime_FromSeconds() usage.

About raised exception, ValueError is raised for Not-A-Number floating pointer number, and OverflowError on conversions (_PyTime_t, time_t, timeval, timespec, etc.).

**Comment by encukou** [comment]
Victor, do you have any intention of putting the removed API back?

**Comment by vstinner** [comment]
> Victor, do you have any intention of putting the removed API back?

My plan in this issue is to add public functions replacing removed _PyTime functions. I'm collecting data to see how the private API was used and which API should be added.

**Comment by vstinner** [comment]
I wrote PR #112135 to add the public functions: PyTime_AsSecondsDouble(), PyTime_GetMonotonicClock(), PyTime_GetPerfCounter() and PyTime_GetSystemClock().

**Comment by scoder** [comment]
Marking this as a release blocker for 3.13 since the API removal is a regression in 3.13.

**Comment by vstinner** [comment]
> Marking this as a release blocker for 3.13 since the API removal is a regression in 3.13.

I reopened my PR https://github.com/python/cpython/pull/112135 and created a C API Working Group issue: https://github.com/capi-workgroup/decisions/issues/8

**Comment by vstinner** [comment]
@scoder: On https://github.com/capi-workgroup/decisions/issues/8 I was asked to change the PyTime API to report errors when reading time. I understand that in practice, it means that the caller must hold the GIL and the function can raise an exception. Would you be fine with such API in `Cython/Includes/cpython/time.pxd`, or do you need a `PyTime_t PyTime_Monotonic()` function which cannot fail and can be used with the GIL released?

If `PyTime_Monotonic()` requires to hold the GIL and can raise an exception, can you modify Cython `cdef inline double time() nogil:` to hold the GIL, catch the exception and return ``0`` on error? Or even modify it to report an error to the caller? I suppose that the second case would be an incompatible change.

**Comment by vstinner** [comment]
@da-woods: Same question for you about Cython.

**Comment by da-woods** [comment]
The exceptions don't represent a big change - when calling `cdef inline double time() nogil` Cython assumes that a return value of `-1` represents a possible exception. Practically that never happens right now, but in principle the signature is of a function that might throw.

Having the GIL is a slightly bigger change. We could make it work and keep the same external interface with a `with gil:` block inside `time()` though (although it wouldn't run *well* in parallel). I think this is something we could live with.

**Comment by encukou** [comment]
New PR that builds on Victor's: https://github.com/python/cpython/pull/115215

**Comment by vstinner** [comment]
@encukou and me added a bare minimum public C API to read time with a `PyTime_t` type (64-bit signed number: nanoseconds): change commit 879f4546bfbc9c47ef228e7c3d2f126f3d8d64bf.

```c
typedef int64_t PyTime_t;
#define PyTime_MIN INT64_MIN
#define PyTime_MAX INT64_MAX

PyAPI_FUNC(double) PyTime_AsSecondsDouble(PyTime_t t);
PyAPI_FUNC(int) PyTime_Monotonic(PyTime_t *result);
PyAPI_FUNC(int) PyTime_PerfCounter(PyTime_t *result);
PyAPI_FUNC(int) PyTime_Time(PyTime_t *result);
```

If someone wants more functions to be exposed, please open a separated issue.

**Comment by vstinner** [comment]
I wrote https://github.com/python/pythoncapi-compat/pull/84 to add these functions to the pythoncapi-compat project. That's how I found PR gh-115713 bug (PyTime_Time() returns 1 instead of 0).

**Comment by vstinner** [comment]
I reused this issue to cleanup the PyTime implementation using new assumptions:

* PyTime_t is always 64-bit
* PyTime_t is always a number of nanoseconds
* I modified code to use the new public functions, rather than old private names.

**Comment by scoder** [comment]
Note that this new implementation does not help Cython because we expose the functions as not requiring the GIL. The new interface requires the GIL because it allows exceptions to be raised.

This leaves us with `_PyTime_TimeUnchecked()`, which is still not a public function.

**Comment by encukou** [comment]
I don't think providing `_PyTime_TimeUnchecked` is in CPython's scope. We do provide platform-specific wrappers, but those come with Python-style error handling.
So far, getting the current time without errors was always possible in practice. But as we add WASI, that's no longer always the case. I don't know if a clock will be required to build CPython, or just some optional parts of it -- the future of WASI and how we handle optional features is still in the cards.

You could use `PyGILState_Ensure` to get the GIL.

**Comment by vstinner** [comment]
> But as we add WASI, that's no longer always the case. I don't know if a clock will be required to build CPython, or just some optional parts of it -- the future of WASI and how we handle optional features is still in the cards.

I'm confident that WASI provides clocks needed by time.time(), time.monotonic() and time.perf_counter(). Otherwise, the problem is not the C API, but that Python will simply not work on such platform. Do you have information saying clearly that WASI has no clock at all?

**Comment by vstinner** [comment]
> I don't think providing _PyTime_TimeUnchecked is in CPython's scope

Python provides many portable C APIs used by Python itself, but also made public for 3rd party C extensions. Example: API for threads and locks. How is time out of scope, but other existing functions are? Python already provides these clocks in the Python time module, just with a different API.

**Comment by vstinner** [comment]
I reopen the issue since Stefan seems to have concerns about Cython.

**Comment by scoder** [comment]
I sketched a different `PyTime_Time()` implementation (without Python exceptions) in https://github.com/capi-workgroup/decisions/issues/8#issuecomment-1959054377

**Comment by vstinner** [comment]
I created https://github.com/capi-workgroup/decisions/issues/15 to propose different options to change PyTime error handling.

**Comment by encukou** [comment]
I don't think this should block the alpha release.

**Comment by encukou** [comment]
Raw time API was added in #118394.
